src/main/java/net/neoforged/neoforge/client/renderstate/RegisterRenderStateModifiersEvent.java

```diff
@@ -0,0 +1,112 @@
+     * @param modifier     The function for modifying the {@link EntityRenderState} and adding custom render data.
+     * @param <E>          The type of the entity
+     * @param <S>          The specific render state type
+     */
+    public <E extends Entity, S extends EntityRenderState> void registerEntityModifier(TypeToken<? extends EntityRenderer<E, S>> baseRenderer, BiConsumer<E, S> modifier) {
+        ensureParametersMatchBounds(baseRenderer);
+        RenderStateExtensions.registerEntity(baseRenderer.getRawType(), modifier);
+    }
+
+    /**
+     * Registers a render state modifier for {@link MapRenderState}s which are run after the texture has been set
+     * and before decorations have been added. Can add custom data to the map using
+     * {@link IRenderStateExtension#setRenderData(ContextKey, Object)}.
+     *
+     * @param modifier The function for modifying the {@link net.minecraft.client.renderer.state.MapRenderState} and adding custom render data.
+     */
+    public void registerMapModifier(BiConsumer<MapItemSavedData, MapRenderState> modifier) {
+        RenderStateExtensions.registerMap(modifier);
+    }
+
+    /**
+     * Registers a render state modifier for {@link MapRenderState.MapDecorationRenderState}s which are run after
+     * vanilla map decoration data has been set. Can add custom data to the map using
+     * {@link IRenderStateExtension#setRenderData(ContextKey, Object)}.
+     *
+     * @param mapDecorationTypeKey Key for the registered {@link MapDecorationType}
+     * @param modifier             The function for modifying the {@link MapRenderState.MapDecorationRenderState} and adding custom render data.
+     */
+    public void registerMapDecorationModifier(ResourceKey<MapDecorationType> mapDecorationTypeKey, MapDecorationRenderStateModifier modifier) {
+        RenderStateExtensions.registerMapDecoration(mapDecorationTypeKey, modifier);
+    }
+
+    private static void ensureParametersMatchBounds(TypeToken<?> baseRenderer) {
+        if (baseRenderer.getType() instanceof ParameterizedType parameterizedType) {
+            Class<?> bound = baseRenderer.getRawType();
+            var parameterized = parameterizedType;
+            do {
+                var userArgs = parameterizedType.getActualTypeArguments();
+                var typeArgs = ((Class) parameterizedType.getRawType()).getTypeParameters();
+                for (int i = 0; i < userArgs.length; i++) {
+                    var userType = userArgs[i];
+                    // ignore wildcards. compilation would catch illegal states, so this must not be an important type to capture
+                    if (userType instanceof WildcardType) continue;
+                    var typeBounds = typeArgs[i].getBounds();
+                    // a generic class with multiple bounds, ie AbstractHoglinRenderer<T extends Mob & HoglinBase>
+                    if (userType instanceof TypeVariable<?> typeVariable) {
```

Comment:
To clarify the issue with this logic, let's say you do 
```java
static class HoglinRendererTypeToken<T extends Mob & HoglinBase> extends TypeToken<AbstractHoglinRenderer<T>> {}
```
And then call 
```java
event.<Zoglin, HoglinRenderState>registerEntityModifier(new HoglinRendererTypeToken<>(), (entity, renderState) -> {
    Zoglin zoglin = entity; // this is safe
});
```
The current verification code will pass this; it is very much not typesafe.